train_gpt2.cu

```diff
@@ -2297,10 +2289,20 @@ void gpt2_backward(GPT2 *model) {
 }
 
 // Compute a mean of a single CPU value across all GPU processes. No-op when multi-GPU is disabled.
-float multi_gpu_cpu_float_mean(float value, const MultiGpuConfig* multi_gpu_config) {
+float multi_gpu_float_mean(float value, float *d_buffer, const MultiGpuConfig* multi_gpu_config) {
 #ifdef MULTI_GPU
-    // MPI doesn't support all reduce with mean, so we sum up, then divide.
+    if (multi_gpu_config->num_processes == 1) return value;
+
     float result;
+
+    if (multi_gpu_config->slurm_managed) {  //If the process is managed by slurm we don't need to use MPI
+        if (d_buffer == NULL) cudaCheck(cudaMalloc(&d_buffer, sizeof(float)));
+        cudaCheck(cudaMemcpy(d_buffer, &value, sizeof(float), cudaMemcpyHostToDevice));
```

Comment:
The reason why I was asking if we could use MPI on Slurm clusters is because of this.

This feels a bit ugly, while a single `MPI_AllReduce` is a lot cleaner.

Reply:
I agree this one looks not great. Also thinking about alternatives.

Reply:
I introduced cuda unified mem buffer with cudaPrefetchAsync. But still internally an additional host to device and device to host memcopy happens. So MPI is cleaner in that sense.
On the other hand I'm thinking about a cleaner way to get rid of MPI and SLURM dependencies by introducing and independent server client interface to manage distributed stuff.